src/Formats/EscapingRuleUtils.h

```diff
@@ -33,5 +34,24 @@ void serializeFieldByEscapingRule(
 void writeStringByEscapingRule(const String & value, WriteBuffer & out, FormatSettings::EscapingRule escaping_rule, const FormatSettings & format_settings);
 
 String readStringByEscapingRule(ReadBuffer & buf, FormatSettings::EscapingRule escaping_rule, const FormatSettings & format_settings);
+String readFieldByEscapingRule(ReadBuffer & buf, FormatSettings::EscapingRule escaping_rule, const FormatSettings & format_settings);
+
+/// Try to determine the type of the field written by a specific escaping rule.
+/// If cannot, return nullptr.
+/// - For Quoted escaping rule we can interpret a single field as a constant
+///   expression and get it's type by evaluation this expression.
+/// - For JSON escaping rule we can use JSON parser to parse a single field
+///   and then convert JSON type of this field to ClickHouse type.
+/// - For CSV escaping rule we can do the next:
+///    - If the field is an unquoted string, then we could try to evaluate it
+///      as a constant expression, and if it fails, treat it as a String.
+///    - If the field is a string in quotes, then we can try to evaluate
+///      expression inside quotes as a constant expression, and if it fails or
+///      the result is a number (we don't parse numbers in quotes) we treat it as a String.
+/// - For TSV and TSVRaw we treat each field as a String (TODO: try to use some tweaks and heuristics here)
```

Comment:
Wouldn't this be too slow?

Reply:
If you are about expression evaluation, then I think yes, it may be slow, but also I think that we shouldn't worry about it, because we process too few lines in schema inference. By the way, I plan to remove evaluating expressions form this place and add small parser for Quoted/CSV/TSV